src/engine/Tableau.cpp

```diff
@@ -1729,27 +1738,41 @@ void Tableau::storeState( TableauState &state ) const
         // Store the merged variables
         state._mergedVariables = _mergedVariables;
     }
+    else
+    {
+        ASSERT( level == TableauStateStorageLevel::STORE_NONE );
+        return;
+    }
 }
 
-void Tableau::restoreState( const TableauState &state )
+void Tableau::restoreState( const TableauState &state,
+                            TableauStateStorageLevel level )
 {
-    if ( _lpSolverType == LPSolverType::GUROBI )
+    if ( level == TableauStateStorageLevel::STORE_BOUNDS_ONLY
+          || _lpSolverType != LPSolverType::NATIVE )
     {
         // Store the bounds
         memcpy( _lowerBounds, state._lowerBounds, sizeof(double) *_n );
         memcpy( _upperBounds, state._upperBounds, sizeof(double) *_n );
 
-        DEBUG({
-                // Restored bounds must be valid. Otherwise, the case split
-                // would not have been performed.
-                for ( unsigned i = 0; i < _n; ++i )
-                    ASSERT( FloatUtils::lte( _lowerBounds[i],
-                                             _upperBounds[i] ) );
-            });
+        _boundsValid = state._boundsValid;
 
-        _boundsValid = true;
+        if ( _lpSolverType == LPSolverType::NATIVE )
+        {
+            // The bounds might be invalid in the state from which we backtrack,
+            // This means we might need to fix the non-basic variable assignments
+            for ( unsigned i = 0; i < _n - _m; ++i )
+            {
+                unsigned variable = _nonBasicIndexToVariable[i];
+                updateVariableToComplyWithLowerBoundUpdate( variable,
```

Comment:
@guykatzz If I don't do this update, the Tableau invariant can be broken (some non-basic variable will go out of bound). Are there better solution than this? 
The alternative would be to not call updateVariableToComplyWithLowerBoundUpdate() when _boundsValid is false in the "tightenLower/UpperBounds methods". But doing that seems to result in invariant being broken elsewhere.
I also tried throw InfeasibleQueryException() in those tightenbounds method once the bounds become invalid. But that is also problematic, because now this exception could be thrown in "popSplit()" method..

Reply:
I guess an alternative would be to store the variables' assignment as part of the stored state? And then have another invariant that says that a stored state is valid, as far as the NB assignments are concerned?

But why do you want to change this? I don't think it's that expensive, performance wise

Reply:
Ah ok. I think I'll keep the current solution.